### ISSUE TITLE
getRoot in afterAttach returns undefined in Jest tests

### ISSUE THREAD

**Issue by garrettg123** [mobxjs/mobx-state-tree#1551]
**_Bug report_**

* [X] I've checked documentation and searched for existing issues
* [X] I've made sure my project is based on the latest MST version
* [X] Fork [this](https://codesandbox.io/s/y64pzxj01) code sandbox or another minimal reproduction. 

**Sandbox link or minimal reproduction code**
https://codesandbox.io/s/mobx-state-tree-todolist-wvupz?file=/index.test.js

**Describe the expected behavior**
`getRoot(self)` should be defined in `afterAttach`

**Describe the observed behavior**
It is undefined

**Comment by mweststrate** [comment]
Looks like a bug indeed! Feel free to check if it reproduces in MST's unit
tests as well, and submit a patch if needed.

**Comment by garrett-gottlieb** [comment]
Hmm ok, I did some digging. Do you think it has something to do here:
https://github.com/mobxjs/mobx-state-tree/blob/5f65318ea2934c44ce27926cfdaae97509801812/packages/mobx-state-tree/src/core/node/object-node.ts#L317
?

**Comment by mweststrate** [comment]
Hard to say, but yes it should hit that piece of code. I'd check out the
repo, create a unit test and attach a debugger. Sorry from the test it
looks like you've hit a bug indeed, but I'm already pretty occupied with
MobX 6, so personally I don't have any cycles on the short term to look
into it. I suspect the reason no-one run into it before is that it is quite
unusual to put a model instance directly in a snapshot like you did,
although it should work. But typically things get instantiated like
`State.create({ model: { foo: 'bar' } })`

**Comment by garrett-gottlieb** [comment]
Ah, I see. In order to spy on the child model's actions, however, is it not necessary to create the model beforehand so you can spy on the actions before attaching it?

**Comment by garrett-gottlieb** [comment]
For reference, here's an example of a simple test where this might be necessary:
```js
    it('loads if not loaded', async () => {
      const Model = types.compose(BaseModel(), types.model())
      const State = types.model({
        model: types.maybe(Model),
      })

      const model = Model.create()
      const loadMock = jest.spyOn(model, 'load').mockResolvedValue()

      // @see https://github.com/mobxjs/mobx-state-tree/issues/1551
      const state = State.create()
      unprotect(state)
      state.model = model

      await loadMock.mock.results[0].value // THIS LINE

      expect(loadMock).toHaveBeenCalledTimes(1)
    })
```
As you can see, on `THIS LINE` the mock needs to be `await`ed in order for the test to finish. In deeper tests, additional assertions might be made, but this is a basic case ensuring that `model.load()` finishes without failing.

**Comment by coolsoftwaretyler** [comment]
Hey @garrettg123 - thanks for all the help in reproducing this bug. I think it's part of a bigger problem documented in a few other issues, and potentially resolved in https://github.com/mobxjs/mobx-state-tree/pull/1952.

We're still waiting some decisions about whether or not this is a breaking change, but if/when it ships, we can come back and test against the work you put together to reproduce.

I'm going to leave this issue open, but mark it as "has PR" for the future.

Thanks!

**Comment by coolsoftwaretyler** [comment]
We merged #1952, so I think this is good to close. Please let me know if you continue to experience this issue on latest MST.
